File: app/src/main/java/org/cerion/stockcharts/watchlist/WatchItemViewModel.java
package org.cerion.stockcharts.watchlist;

import android.os.Handler;

import androidx.databinding.ObservableField;

import org.cerion.stockcharts.common.Utils;
import org.cerion.stocks.core.PriceList;
import org.cerion.stocks.core.charts.StockChart;
import org.cerion.stocks.core.functions.conditions.ICondition;
import org.cerion.stocks.core.web.DataAPI;

public class WatchItemViewModel {

    private DataAPI api;
    private ICondition condition;
    private String symbol;
    private Handler handler;

    public final ObservableField<String> price = new ObservableField<>("--");
    public final ObservableField<String> change = new ObservableField<>("--");
    public final ObservableField<Integer> weekPosition = new ObservableField<>(0);
    public final ObservableField<Integer> yearPosition = new ObservableField<>(0);
    public final ObservableField<Boolean> isTrue = new ObservableField<>(false);
    public final ObservableField<Boolean> loading = new ObservableField<>(true);

    public WatchItemViewModel(Handler handler, DataAPI api, ICondition condition, String symbol) {
        this.handler = handler;
        this.api = api;
        this.condition = condition;
        this.symbol = symbol;
    }

    public void load() {
        /*
        handler.post(new Runnable() {
            @Override
            public void run() {
                loading.set(true);
                try {
                    PriceList prices = new PriceList(getSymbol(), api.getPrices(getSymbol(), Interval.DAILY, Constants.START_DATE_DAILY));
                    apply(prices);
                } catch (Exception e) {
                    // TODO set error state
                    e.printStackTrace();
                }
                loading.set(false);
            }
        });

         */
    }

    public String getSymbol() {
        return symbol;
    }

    public String getCondition() {
        return condition.toString();
    }

    public StockChart getChart() {
        return condition.getChart();
    }

    private void apply(PriceList list) {

        int size = list.size();
        float price = list.getLast().getClose();
        float change = list.getLast().getPercentDiff(list.get(size - 2));

        // TODO include current quote since its not in pricelist
        float low = list.getLow().get(size - 1);
        for(int j = size - 5; j < size; j++) {
            if (list.getLow().get(j) < low)
                low = list.getLow().get(j);
        }

        float high = list.getHigh().get(size - 1);
        for(int j = size - 5; j < size; j++) {
            if (list.getHigh().get(j) > high)
                high = list.getHigh().get(j);
        }

        float range = high - low;
        float diff = list.getLast().getClose() - low;
        float percent = diff / range;
        weekPosition.set((int)(percent * 100));

        // TODO add function to get highest high in given range to PriceList

        // Year
        int start = 250;
        if (list.size() < start)
            start = 0;

        low = list.getLow().get(size - 1);
        for(int j = size - start; j < size; j++) {
            if (list.getLow().get(j) < low)
                low = list.getLow().get(j);
        }

        high = list.getHigh().get(size - 1);
        for(int j = size - start; j < size; j++) {
            if (list.getHigh().get(j) > high)
                high = list.getHigh().get(j);
        }

        range = high - low;
        diff = list.getLast().getClose() - low;
        percent = diff / range;
        yearPosition.set((int)(percent * 100));

        isTrue.set(condition.eval(list));

        // Set strings
        this.price.set(Utils.decimalFormat.format(price));
        this.change.set(Utils.decimalFormat.format(change) + "%");
    }



}

File: app/src/main/java/org/cerion/stockcharts/ui/watchlist/WatchItemViewModel.kt
package org.cerion.stockcharts.ui.watchlist

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import org.cerion.marketdata.core.charts.StockChart
import org.cerion.marketdata.core.functions.conditions.ICondition
import org.cerion.marketdata.core.model.Interval
import org.cerion.marketdata.core.model.OHLCVTable
import org.cerion.stockcharts.repository.CachedPriceListRepository

class WatchItemViewModel(private val repo: CachedPriceListRepository, private val condition: ICondition, val symbol: String) {

    // TODO change to livedata
    //val price = ObservableField("--")
    //val change = ObservableField("--")
    //val weekPosition = ObservableField(0)
    //val yearPosition = ObservableField(0)
    //val isTrue = ObservableField(false)

    private val _loading = MutableLiveData(false)
    val loading: LiveData<Boolean>
        get() = _loading

    suspend fun update() {
        withContext(Dispatchers.IO) {
            delay(1000)

            _loading.value = true
            try {
                val table = repo.get(symbol, Interval.DAILY)
                apply(table)
            } catch (e: Exception) {
                // TODO set error state
                e.printStackTrace()
            }

            _loading.value = false
        }
    }

    fun getCondition(): String {
        return condition.toString()
    }

    val chart: StockChart
        get() = condition.chart

    private fun apply(table: OHLCVTable) {
        /*
        val size = table.size
        val price = table.last().close
        val change = table.last().getPercentDiff(table[size - 2])

        // TODO include current quote since its not in pricelist
        var low = table.low[size - 1]
        for (j in size - 5 until size) {
            if (table.low[j] < low) low = table.low[j]
        }

        var high = table.high[size - 1]
        for (j in size - 5 until size) {
            if (table.high[j] > high) high = table.high[j]
        }

        var range = high - low
        var diff = table.last().close - low
        var percent = diff / range
        weekPosition.set((percent * 100).toInt())

        // TODO add function to get highest high in given range to PriceList

        // Year
        var start = 250
        if (table.size < start) start = 0
        low = table.low[size - 1]
        for (j in size - start until size) {
            if (table.low[j] < low) low = table.low[j]
        }

        high = table.high[size - 1]
        for (j in size - start until size) {
            if (table.high[j] > high) high = table.high[j]
        }

        range = high - low
        diff = table.last().close - low
        percent = diff / range

        yearPosition.set((percent * 100).toInt())
        isTrue.set(condition.eval(table))

        // Set strings
        this.price.set(Utils.decimalFormat.format(price.toDouble()))
        this.change.set(Utils.decimalFormat.format(change.toDouble()) + "%")

         */
    }
}